File: app/src/main/java/ru/melodin/fast/api/model/VKConversation.java
package ru.melodin.fast.api.model;

import android.text.TextUtils;

import androidx.annotation.NonNull;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.Serializable;
import java.util.ArrayList;

public class VKConversation extends VKModel implements Serializable {

    private static final long serialVersionUID = 1L;

    public static int count;
    public static ArrayList<VKUser> users = new ArrayList<>();
    public static ArrayList<VKGroup> groups = new ArrayList<>();

    private int peerId;
    private int localId;
    private int readIn;
    private int readOut;
    private int lastMessageId;
    private int unread;
    private int membersCount;

    private boolean canWrite;
    private int reason;

    private boolean read;
    private boolean groupChannel;

    private VKMessage pinned, last;

    private String title;
    private Type type;
    private State state = State.IN;
    private String photo50, photo100, photo200;

    //acl
    private boolean canChangePin, canChangeInfo, canChangeInviteLink, canInvite, canPromoteUsers, canSeeInviteLink;

    //push_settings
    private int disabledUntil;
    private boolean disabledForever;
    private boolean noSound;

    public enum State {
        IN, KICKED, LEFT
    }

    public enum Type {
        CHAT, GROUP, USER
    }

    public enum Reason {
        KICKED, LEFT, USER_DELETED, USER_BLACKLIST, USER_PRIVACY, MESSAGES_OFF, MESSAGES_BLOCKED, NO_ACCESS_CHAT, NO_ACCESS_EMAIL, NO_ACCESS_GROUP, UNKNOWN
    }

    @NonNull
    @Override
    public String toString() {
        return title;
    }

    public VKConversation() {

    }

    public static State getState(String state) {
        if (TextUtils.isEmpty(state)) return null;
        switch (state) {
            case "kicked":
                return State.KICKED;
            case "in":
                return State.IN;
            case "left":
                return State.LEFT;
        }

        return null;
    }

    public static String getState(State state) {
        if (state == null) return null;
        switch (state) {
            case KICKED:
                return "kicked";
            case LEFT:
                return "left";
            case IN:
                return "in";
        }

        return null;
    }

    public static int getReason(Reason reason) {
        switch (reason) {
            case USER_BLACKLIST:
                return 900;
            case USER_PRIVACY:
                return 902;
            case USER_DELETED:
                return 18;
            case LEFT:
                return 2;
            case KICKED:
                return 1;
            case MESSAGES_OFF:
                return 915;
            case MESSAGES_BLOCKED:
                return 916;
            case NO_ACCESS_CHAT:
                return 917;
            case NO_ACCESS_EMAIL:
                return 918;
            case NO_ACCESS_GROUP:
                return 203;
            default:
                return -1;
        }
    }

    public static Reason getReason(int reason) {
        switch (reason) {
            case 900:
                return Reason.USER_BLACKLIST;
            case 902:
                return Reason.USER_PRIVACY;
            case 18:
                return Reason.USER_DELETED;
            case 2:
                return Reason.LEFT;
            case 1:
                return Reason.KICKED;
            case 915:
                return Reason.MESSAGES_OFF;
            case 916:
                return Reason.MESSAGES_BLOCKED;
            case 917:
                return Reason.NO_ACCESS_CHAT;
            case 918:
                return Reason.NO_ACCESS_EMAIL;
            case 203:
                return Reason.NO_ACCESS_GROUP;
            default:
                return Reason.UNKNOWN;
        }
    }

    public VKConversation(JSONObject o, JSONObject msg) throws JSONException {

        if (msg != null)
            last = new VKMessage(msg);

        JSONObject peer = o.optJSONObject("peer");
        this.peerId = peer.optInt("id", -1);
        this.localId = peer.optInt("local_id", -1);

        String type = peer.optString("type");
        this.type = getType(type);

        this.readIn = o.optInt("in_read");
        this.readOut = o.optInt("out_read");
        this.lastMessageId = o.optInt("last_message_id");
        this.unread = o.optInt("unread_count");

        this.read = this.last == null || (this.last.isOut() && this.readOut == this.lastMessageId || !this.last.isOut() && this.readIn == this.lastMessageId);

        JSONObject canWrite = o.optJSONObject("can_write");
        this.canWrite = canWrite.optBoolean("allowed");
        if (!this.canWrite) {
            this.reason = canWrite.optInt("reason", -1);
        }

        JSONObject push_settings = o.optJSONObject("push_settings");
        if (push_settings != null) {
            this.disabledUntil = push_settings.optInt("disabled_until", -1);
            this.disabledForever = push_settings.optBoolean("disabled_forever", false);
            this.noSound = push_settings.optBoolean("no_sound");
        }

        JSONObject ch = o.optJSONObject("chat_settings");
        if (ch != null) {
            this.title = ch.optString("title");
            this.membersCount = ch.optInt("members_count");
            this.state = VKConversation.getState(ch.optString("state"));
            this.groupChannel = ch.optBoolean("is_group_channel");

            JSONObject photo = ch.optJSONObject("photo");
            if (photo != null) {
                this.photo50 = photo.optString("photo_50");
                this.photo100 = photo.optString("photo_100");
                this.photo200 = photo.optString("photo_200");
            }

            JSONObject acl = ch.optJSONObject("acl");
            if (acl != null) {
                this.canInvite = acl.optBoolean("can_invite");
                this.canPromoteUsers = acl.optBoolean("can_promote_users");
                this.canSeeInviteLink = acl.optBoolean("can_see_invite_link");
                this.canChangeInviteLink = acl.optBoolean("can_change_invite_link");
                this.canChangeInfo = acl.optBoolean("can_change_info");
                this.canChangePin = acl.optBoolean("can_change_pin");
            }

            JSONObject pinned = ch.optJSONObject("pinned_message");
            if (pinned != null) {
                this.pinned = new VKMessage(pinned);
            }
        }
    }

    public boolean isGroupChannel() {
        return groupChannel;
    }

    public void setGroupChannel(boolean groupChannel) {
        this.groupChannel = groupChannel;
    }

    public Type getType() {
        return type;
    }

    public static Type getType(int peerId) {
        if (VKConversation.isChatId(peerId)) return Type.CHAT;
        if (VKGroup.isGroupId(peerId)) return Type.GROUP;
        return Type.USER;
    }

    public static Type getType(String type) {
        if (TextUtils.isEmpty(type)) return null;
        switch (type) {
            case "user":
                return Type.USER;
            case "group":
                return Type.GROUP;
            case "chat":
                return Type.CHAT;
            default:
                return null;
        }
    }

    public static String getType(Type type) {
        if (type == null) return null;
        switch (type) {
            case USER:
                return "user";
            case GROUP:
                return "group";
            case CHAT:
                return "chat";
            default:
                return null;
        }
    }

    private static boolean isChatId(int peerId) {
        return peerId > 2_000_000_00;
    }

    public boolean isChat() {
        return getType() == Type.CHAT && !isGroupChannel();
    }

    public boolean isFromGroup() {
        return VKGroup.isGroupId(this.last.getFromId());
    }

    public boolean isGroup() {
        return getType() == Type.GROUP && !isGroupChannel();
    }

    public boolean isUser() {
        return !isGroup() && !isChat() && !isGroupChannel();
    }

    public boolean isFromUser() {
        return !isFromGroup();
    }

    public boolean isNotificationsDisabled() {
        return this.disabledForever || this.disabledUntil > 0 || this.noSound;
    }


    public int getPeerId() {
        return peerId;
    }

    public int getLocalId() {
        return localId;
    }

    public void setPeerId(int peerId) {
        this.peerId = peerId;
    }

    public void setLocalId(int localId) {
        this.localId = localId;
    }

    public int getReadIn() {
        return readIn;
    }

    public void setReadIn(int readIn) {
        this.readIn = readIn;
    }

    public int getReadOut() {
        return readOut;
    }

    public void setReadOut(int readOut) {
        this.readOut = readOut;
    }

    public int getLastMessageId() {
        return lastMessageId;
    }

    public void setLastMessageId(int lastMessageId) {
        this.lastMessageId = lastMessageId;
    }

    public int getUnread() {
        return unread;
    }

    public void setUnread(int unread) {
        this.unread = unread;
    }

    public int getMembersCount() {
        return membersCount;
    }

    public void setMembersCount(int membersCount) {
        this.membersCount = membersCount;
    }

    public boolean isCanWrite() {
        return canWrite;
    }

    public void setCanWrite(boolean canWrite) {
        this.canWrite = canWrite;
    }

    public int getReason() {
        return reason;
    }

    public void setReason(int reason) {
        this.reason = reason;
    }

    public boolean isRead() {
        return read;
    }

    public void setRead(boolean read) {
        this.read = read;
    }

    public VKMessage getPinned() {
        return pinned;
    }

    public void setPinned(VKMessage pinned) {
        this.pinned = pinned;
    }

    public VKMessage getLast() {
        return last;
    }

    public void setLast(VKMessage last) {
        this.last = last;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public void setType(Type type) {
        this.type = type;
    }

    public State getState() {
        return state;
    }

    public void setState(State state) {
        this.state = state;
    }

    public String getPhoto50() {
        return photo50;
    }

    public void setPhoto50(String photo50) {
        this.photo50 = photo50;
    }

    public String getPhoto100() {
        return photo100;
    }

    public void setPhoto100(String photo100) {
        this.photo100 = photo100;
    }

    public String getPhoto200() {
        return photo200;
    }

    public void setPhoto200(String photo200) {
        this.photo200 = photo200;
    }

    public boolean isCanChangePin() {
        return canChangePin;
    }

    public void setCanChangePin(boolean canChangePin) {
        this.canChangePin = canChangePin;
    }

    public boolean isCanChangeInfo() {
        return canChangeInfo;
    }

    public void setCanChangeInfo(boolean canChangeInfo) {
        this.canChangeInfo = canChangeInfo;
    }

    public boolean isCanChangeInviteLink() {
        return canChangeInviteLink;
    }

    public void setCanChangeInviteLink(boolean canChangeInviteLink) {
        this.canChangeInviteLink = canChangeInviteLink;
    }

    public boolean isCanInvite() {
        return canInvite;
    }

    public void setCanInvite(boolean canInvite) {
        this.canInvite = canInvite;
    }

    public boolean isCanPromoteUsers() {
        return canPromoteUsers;
    }

    public void setCanPromoteUsers(boolean canPromoteUsers) {
        this.canPromoteUsers = canPromoteUsers;
    }

    public boolean isCanSeeInviteLink() {
        return canSeeInviteLink;
    }

    public void setCanSeeInviteLink(boolean canSeeInviteLink) {
        this.canSeeInviteLink = canSeeInviteLink;
    }

    public int getDisabledUntil() {
        return disabledUntil;
    }

    public void setDisabledUntil(int disabledUntil) {
        this.disabledUntil = disabledUntil;
    }

    public boolean isDisabledForever() {
        return disabledForever;
    }

    public void setDisabledForever(boolean disabledForever) {
        this.disabledForever = disabledForever;
    }

    public boolean isNoSound() {
        return noSound;
    }

    public void setNoSound(boolean noSound) {
        this.noSound = noSound;
    }
}

File: app/src/main/java/ru/melodin/fast/api/model/VKConversation.kt
package ru.melodin.fast.api.model

import android.util.Log
import org.greenrobot.eventbus.EventBus
import org.jetbrains.annotations.Contract
import org.json.JSONException
import org.json.JSONObject
import ru.melodin.fast.database.CacheStorage
import ru.melodin.fast.database.MemoryCache
import ru.melodin.fast.util.Keys
import java.io.Serializable
import java.util.*

class VKConversation : VKModel, Serializable {

    var peerId = 0
    var ownerId = 0
    var localId = 0
    var readIn = 0
    var readOut = 0
    var lastMessageId = 0
    var unread = 0
    var membersCount = 0

    var isCanWrite: Boolean = false
    var reason = 0

    var isRead: Boolean = false
    var isGroupChannel: Boolean = false

    var pinned: VKMessage? = null
    var lastMessage: VKMessage? = null

    var title: String? = null
    var type: Type? = null
    var state: State? = State.IN
    var photo50: String? = null
    var photo100: String? = null
    var photo200: String? = null

    //acl
    var isCanChangePin: Boolean = false
    var isCanChangeInfo: Boolean = false
    var isCanChangeInviteLink: Boolean = false
    var isCanInvite: Boolean = false
    var isCanPromoteUsers: Boolean = false
    var isCanSeeInviteLink: Boolean = false
    var isCanModerate: Boolean = false

    //push_settings
    var disabledUntil: Int = 0
    var isDisabledForever: Boolean = false
    var isNoSound: Boolean = false

    val fullTitle: String?
        get() {
            val peerId = peerId

            val group: VKGroup?
            val user: VKUser?

            when (type) {
                Type.GROUP -> {
                    group = CacheStorage.getGroup(VKGroup.toGroupId(peerId))
                    return if (group == null) {
                        EventBus.getDefault()
                            .postSticky(arrayOf(Keys.NEED_LOAD_ID, peerId, javaClass.simpleName))
                        null
                    } else {
                        group.toString()
                    }
                }
                Type.USER -> {
                    user = CacheStorage.getUser(peerId)
                    return if (user == null) {
                        EventBus.getDefault()
                            .postSticky(arrayOf(Keys.NEED_LOAD_ID, peerId, javaClass.simpleName))
                        null
                    } else {
                        user.toString()
                    }
                }
                Type.CHAT -> return if (isGroupChannel) {
                    group = CacheStorage.getGroup(VKGroup.toGroupId(ownerId))
                    if (group == null) {
                        EventBus.getDefault()
                            .postSticky(arrayOf(Keys.NEED_LOAD_ID, peerId, javaClass.simpleName))
                        null
                    } else {
                        group.toString()
                    }
                } else {
                    toString()
                }
            }

            return null
        }

    val photo: String?
        get() {
            val peerId = peerId

            val group: VKGroup?
            val user: VKUser?

            when (type) {
                Type.GROUP -> {
                    group = MemoryCache.getGroup(VKGroup.toGroupId(peerId))
                    return if (group == null) {
                        EventBus.getDefault()
                            .postSticky(arrayOf(Keys.NEED_LOAD_ID, peerId, javaClass.simpleName))
                        null
                    } else {
                        group.photo200
                    }
                }
                Type.USER -> {
                    user = MemoryCache.getUser(peerId)
                    return if (user == null) {
                        EventBus.getDefault()
                            .postSticky(arrayOf(Keys.NEED_LOAD_ID, peerId, javaClass.simpleName))
                        null
                    } else {
                        user.photo200
                    }
                }
                Type.CHAT -> return if (isGroupChannel) {
                    group = MemoryCache.getGroup(VKGroup.toGroupId(ownerId))
                    if (group == null) {
                        EventBus.getDefault()
                            .postSticky(arrayOf(Keys.NEED_LOAD_ID, peerId, javaClass.simpleName))
                        null
                    } else {
                        group.photo200
                    }
                } else {
                    photo200
                }
            }

            return null
        }

    val isChat: Boolean
        get() = type == Type.CHAT && !isGroupChannel

    val isFromGroup: Boolean
        get() = VKGroup.isGroupId(this.lastMessage!!.fromId)

    val isGroup: Boolean
        get() = type == Type.GROUP && !isGroupChannel

    val isFromUser: Boolean
        get() = !isFromGroup

    val isNotificationsDisabled: Boolean
        get() = this.isDisabledForever || this.disabledUntil > 0 || this.isNoSound

    constructor()

    @Throws(JSONException::class)
    constructor(o: JSONObject, msg: JSONObject?) {
        Log.d("FVKConversation", o.toString())

        if (msg != null)
            lastMessage = VKMessage(msg)

        val peer = o.optJSONObject("peer")
        this.peerId = peer!!.optInt("id", -1)
        this.localId = peer.optInt("local_id", -1)

        val type = peer.optString("type")
        this.type = getType(type)

        this.readIn = o.optInt("in_read")
        this.readOut = o.optInt("out_read")
        this.lastMessageId = o.optInt("last_message_id", -1)
        this.unread = o.optInt("unread_count")

        this.isRead =
            this.lastMessage == null || this.lastMessage!!.isOut && this.readOut == this.lastMessageId || !this.lastMessage!!.isOut && this.readIn == this.lastMessageId

        val canWrite = o.optJSONObject("can_write")
        this.isCanWrite = canWrite!!.optBoolean("allowed")
        if (!this.isCanWrite) {
            this.reason = canWrite.optInt("reason", -1)
        }

        val pushSettings = o.optJSONObject("push_settings")
        if (pushSettings != null) {
            this.disabledUntil = pushSettings.optInt("disabled_until", -1)
            this.isDisabledForever = pushSettings.optBoolean("disabled_forever", false)
            this.isNoSound = pushSettings.optBoolean("no_sound")
        }

        val ch = o.optJSONObject("chat_settings")
        if (ch != null) {
            this.ownerId = ch.optInt("owner_id", -1)
            this.title = ch.optString("title")
            this.membersCount = ch.optInt("members_count")
            this.state = getState(ch.optString("state"))
            this.isGroupChannel = ch.optBoolean("is_group_channel")

            val photo = ch.optJSONObject("photo")
            if (photo != null) {
                this.photo50 = photo.optString("photo_50")
                this.photo100 = photo.optString("photo_100")
                this.photo200 = photo.optString("photo_200")
            }

            val acl = ch.optJSONObject("acl")
            if (acl != null) {
                this.isCanInvite = acl.optBoolean("can_invite")
                this.isCanPromoteUsers = acl.optBoolean("can_promote_users")
                this.isCanSeeInviteLink = acl.optBoolean("can_see_invite_link")
                this.isCanChangeInviteLink = acl.optBoolean("can_change_invite_link")
                this.isCanChangeInfo = acl.optBoolean("can_change_info")
                this.isCanChangePin = acl.optBoolean("can_change_pin")
                this.isCanModerate = acl.optBoolean("can_moderate")
            }

            val pinned = ch.optJSONObject("pinned_message")
            if (pinned != null) {
                this.pinned = VKMessage(pinned)
            }
        }
    }

    override fun toString(): String {
        return title ?: ""
    }

    enum class State {
        IN, KICKED, LEFT
    }

    enum class Type {
        CHAT, GROUP, USER
    }

    enum class Reason {
        KICKED, LEFT, USER_DELETED, USER_BLACKLIST, USER_PRIVACY, MESSAGES_OFF, MESSAGES_BLOCKED, NO_ACCESS_CHAT, NO_ACCESS_EMAIL, NO_ACCESS_GROUP, UNKNOWN
    }

    companion object {

        private const val serialVersionUID = 1L

        var count: Int = 0
        var users = ArrayList<VKUser>()
        var groups = ArrayList<VKGroup>()

        fun getState(state: String?): State? {
            return when (state) {
                null -> null
                "kicked" -> return State.KICKED
                "in" -> return State.IN
                "left" -> return State.LEFT
                else -> null
            }
        }

        fun getState(state: State?): String? {
            if (state == null) return null
            return when (state) {
                State.KICKED -> "kicked"
                State.LEFT -> "left"
                State.IN -> "in"
            }
        }

        fun getReason(reason: Reason): Int {
            return when (reason) {
                Reason.USER_BLACKLIST -> 900
                Reason.USER_PRIVACY -> 902
                Reason.USER_DELETED -> 18
                Reason.LEFT -> 2
                Reason.KICKED -> 1
                Reason.MESSAGES_OFF -> 915
                Reason.MESSAGES_BLOCKED -> 916
                Reason.NO_ACCESS_CHAT -> 917
                Reason.NO_ACCESS_EMAIL -> 918
                Reason.NO_ACCESS_GROUP -> 203
                else -> -1
            }
        }

        fun getReason(reason: Int): Reason {
            return when (reason) {
                900 -> Reason.USER_BLACKLIST
                902 -> Reason.USER_PRIVACY
                18 -> Reason.USER_DELETED
                2 -> Reason.LEFT
                1 -> Reason.KICKED
                915 -> Reason.MESSAGES_OFF
                916 -> Reason.MESSAGES_BLOCKED
                917 -> Reason.NO_ACCESS_CHAT
                918 -> Reason.NO_ACCESS_EMAIL
                203 -> Reason.NO_ACCESS_GROUP
                else -> Reason.UNKNOWN
            }
        }

        fun getType(peerId: Int): Type {
            if (isChatId(peerId)) return Type.CHAT
            return if (VKGroup.isGroupId(peerId)) Type.GROUP else Type.USER
        }

        fun getType(type: String?): Type? {
            return when (type) {
                null -> null
                "user" -> Type.USER
                "group" -> Type.GROUP
                "chat" -> Type.CHAT
                else -> null
            }
        }

        @Contract(value = "null -> null", pure = true)
        fun getType(type: Type?): String? {
            return when (type) {
                null -> null
                Type.USER -> "user"
                Type.GROUP -> "group"
                Type.CHAT -> "chat"
            }
        }

        @Contract(pure = true)
        fun isChatId(peerId: Int): Boolean {
            return peerId > 2000000000
        }

        fun toChatId(id: Int): Int {
            return if (id > 2000000000) id - 2000000000 else id
        }
    }
}
